{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "e34af545-f96d-487f-8e12-7a556074323d",
   "metadata": {},
   "outputs": [],
   "source": [
    "library('Seurat')\n",
    "library('SeuratDisk')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "0e734651-70aa-4b9b-bab8-3c3cc5a9aef1",
   "metadata": {},
   "outputs": [],
   "source": [
    "rds_data <- readRDS(\"C:/Users/nidhi/Downloads/ependymal_cells.rds\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "c4f09739-e080-4531-8934-1faf64fb754b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Formal class 'seurat' [package \"Seurat\"] with 0 slots\n",
      " Named list()\n",
      "  ..$ raw.data    :Formal class 'dgCMatrix' [package \"Matrix\"] with 6 slots\n",
      "  .. .. ..@ i       : int [1:4064873] 0 21 30 31 51 63 75 78 84 91 ...\n",
      "  .. .. ..@ p       : int [1:3063] 0 1210 2655 4862 5660 6111 7319 8287 9656 10911 ...\n",
      "  .. .. ..@ Dim     : int [1:2] 17138 3062\n",
      "  .. .. ..@ Dimnames:List of 2\n",
      "  .. .. .. ..$ : chr [1:17138] \"X.343C11.2\" \"X0610007P14Rik\" \"X0610009B22Rik\" \"X0610009E02Rik\" ...\n",
      "  .. .. .. ..$ : chr [1:3062] \"DRF1.1_bcGRPK\" \"DRF1.1_bcCUSZ\" \"DRF1.1_bcCBVN\" \"DRF1.1_bcALFX\" ...\n",
      "  .. .. ..@ x       : num [1:4064873] 1 1 4 2 2 1 1 1 1 3 ...\n",
      "  .. .. ..@ factors : list()\n",
      "  ..$ data        :Formal class 'dgCMatrix' [package \"Matrix\"] with 6 slots\n",
      "  .. .. ..@ i       : int [1:4064873] 0 21 30 31 51 63 75 78 84 91 ...\n",
      "  .. .. ..@ p       : int [1:3063] 0 1210 2655 4862 5660 6111 7319 8287 9656 10911 ...\n",
      "  .. .. ..@ Dim     : int [1:2] 17138 3062\n",
      "  .. .. ..@ Dimnames:List of 2\n",
      "  .. .. .. ..$ : chr [1:17138] \"X.343C11.2\" \"X0610007P14Rik\" \"X0610009B22Rik\" \"X0610009E02Rik\" ...\n",
      "  .. .. .. ..$ : chr [1:3062] \"DRF1.1_bcGRPK\" \"DRF1.1_bcCUSZ\" \"DRF1.1_bcCBVN\" \"DRF1.1_bcALFX\" ...\n",
      "  .. .. ..@ x       : num [1:4064873] 1.56 1.56 2.77 2.14 2.14 ...\n",
      "  .. .. ..@ factors : list()\n",
      "  ..$ scale.data  : num [1:17138, 1:3062] 5.7397 -0.4222 -0.2287 -0.0776 -0.1566 ...\n",
      "  .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. ..$ : chr [1:17138] \"X.343C11.2\" \"X0610007P14Rik\" \"X0610009B22Rik\" \"X0610009E02Rik\" ...\n",
      "  .. .. ..$ : chr [1:3062] \"DRF1.1_bcGRPK\" \"DRF1.1_bcCUSZ\" \"DRF1.1_bcCBVN\" \"DRF1.1_bcALFX\" ...\n",
      "  ..$ var.genes   : chr [1:3340] \"Hbb.bs\" \"Cartpt\" \"Plp1\" \"Meg3\" ...\n",
      "  ..$ is.expr     : num 0\n",
      "  ..$ ident       : Ord.factor w/ 1 level \"Ependymal Cells\": 1 1 1 1 1 1 1 1 1 1 ...\n",
      "  .. ..- attr(*, \"names\")= chr [1:3062] \"DRF1.1_bcGRPK\" \"DRF1.1_bcCUSZ\" \"DRF1.1_bcCBVN\" \"DRF1.1_bcALFX\" ...\n",
      "  ..$ meta.data   :'data.frame':\t3062 obs. of  20 variables:\n",
      "  .. ..$ nGene                                : num [1:3062] 1210 1445 2207 798 451 ...\n",
      "  .. ..$ nUMI                                 : num [1:3062] 2668 3350 6601 1977 1003 ...\n",
      "  .. ..$ Reads                                : int [1:3062] 52997 69749 189948 49538 23634 58251 41760 53343 51998 35611 ...\n",
      "  .. ..$ UMIFM                                : int [1:3062] 2951 3725 7974 2192 1085 3176 2180 3616 3392 2555 ...\n",
      "  .. ..$ BatchID                              : int [1:3062] 1 1 1 1 1 1 1 1 1 1 ...\n",
      "  .. ..$ Condition                            : Factor w/ 1 level \"Control\": 1 1 1 1 1 1 1 1 1 1 ...\n",
      "  .. ..$ orig.ident                           : chr [1:3062] \"DRF1.1\" \"DRF1.1\" \"DRF1.1\" \"DRF1.1\" ...\n",
      "  .. ..$ Sex                                  : Factor w/ 2 levels \"Female\",\"Male\": 1 1 1 1 1 1 1 1 1 1 ...\n",
      "  .. ..$ percent.mito                         : num [1:3062] 0.0161 0.026 0.0147 0.0445 0.0828 ...\n",
      "  .. ..$ var.ratio.pca                        : num [1:3062] 0.97 1.005 0.939 0.88 0.908 ...\n",
      "  .. ..$ res.3                                : chr [1:3062] \"6\" \"6\" \"6\" \"6\" ...\n",
      "  .. ..$ ClusterNumbers_cca.aligned_1.35_res_3: chr [1:3062] \"6\" \"6\" \"6\" \"6\" ...\n",
      "  .. ..$ res.2                                : chr [1:3062] \"5\" \"5\" \"5\" \"5\" ...\n",
      "  .. ..$ ClusterNumbers_cca.aligned_1.35_res_2: chr [1:3062] \"5\" \"5\" \"5\" \"5\" ...\n",
      "  .. ..$ tree.ident                           : int [1:3062] 21 21 21 21 21 21 21 21 21 21 ...\n",
      "  .. ..$ ClusterTreeNumbers_filtered          : chr [1:3062] \"21\" \"21\" \"21\" \"21\" ...\n",
      "  .. ..$ CellClassNames_filtered              : chr [1:3062] \"Ependymal Cells\" \"Ependymal Cells\" \"Ependymal Cells\" \"Ependymal Cells\" ...\n",
      "  .. ..$ Curated_subtypeLabels                : chr [1:3062] \"Ependymal_Cdhr4_Cd36_Pdyn_Aq_avm\" \"Ependymal_Cdhr4_Cd36_Pdyn_Aq_avm\" \"Ependymal_Cdhr4_Cd36_Pdyn_Aq_avm\" \"Ependymal_Cdhr4_Cd36_Pdyn_Aq_avm\" ...\n",
      "  .. ..$ Curated_cellTypeLabels               : chr [1:3062] \"Ependymal Cells\" \"Ependymal Cells\" \"Ependymal Cells\" \"Ependymal Cells\" ...\n",
      "  .. ..$ Curated_cellTypeLabels_ordered       : chr [1:3062] \"Ependymal Cells\" \"Ependymal Cells\" \"Ependymal Cells\" \"Ependymal Cells\" ...\n",
      "  ..$ project.name: chr \"DRNCellTypes\"\n",
      "  ..$ dr          :List of 3\n",
      "  .. ..$ cca        :Formal class 'dim.reduction' [package \"Seurat\"] with 0 slots\n",
      " Named list()\n",
      "  .. .. .. ..$ cell.embeddings   : num [1:3062, 1:50] 0.000939 0.000543 0.001048 -0.000199 0.001333 ...\n",
      "  .. .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. .. ..$ : chr [1:3062] \"DRF1.1_bcGRPK\" \"DRF1.1_bcCUSZ\" \"DRF1.1_bcCBVN\" \"DRF1.1_bcALFX\" ...\n",
      "  .. .. .. .. .. ..$ : chr [1:50] \"CC1\" \"CC2\" \"CC3\" \"CC4\" ...\n",
      "  .. .. .. ..$ gene.loadings     : num [1:3320, 1:50] 3.65 7.57 64.22 59.88 32.47 ...\n",
      "  .. .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. .. ..$ : chr [1:3320] \"Hbb.bs\" \"Cartpt\" \"Plp1\" \"Meg3\" ...\n",
      "  .. .. .. .. .. ..$ : chr [1:50] \"CC1\" \"CC2\" \"CC3\" \"CC4\" ...\n",
      "  .. .. .. ..$ gene.loadings.full: num [1:17138, 1:50] 1.81 -30.39 -10.89 3.12 -5.39 ...\n",
      "  .. .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. .. ..$ : chr [1:17138] \"X.343C11.2\" \"X0610007P14Rik\" \"X0610009B22Rik\" \"X0610009E02Rik\" ...\n",
      "  .. .. .. .. .. ..$ : chr [1:50] \"CC1\" \"CC2\" \"CC3\" \"CC4\" ...\n",
      "  .. .. .. ..$ sdev              : num(0) \n",
      "  .. .. .. ..$ key               : chr \"CC\"\n",
      "  .. .. .. ..$ jackstraw         : symbol \u0001NULL\u0001\n",
      "  .. .. .. ..$ misc              : symbol \u0001NULL\u0001\n",
      "  .. ..$ cca.aligned:Formal class 'dim.reduction' [package \"Seurat\"] with 0 slots\n",
      " Named list()\n",
      "  .. .. .. ..$ cell.embeddings   : num [1:3062, 1:35] 0.3 0.238 0.317 0.12 0.362 ...\n",
      "  .. .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. .. ..$ : chr [1:3062] \"DRF1.1_bcGRPK\" \"DRF1.1_bcCUSZ\" \"DRF1.1_bcCBVN\" \"DRF1.1_bcALFX\" ...\n",
      "  .. .. .. .. .. ..$ : chr [1:35] \"ACC1\" \"ACC2\" \"ACC3\" \"ACC4\" ...\n",
      "  .. .. .. ..$ gene.loadings     : num[0 , 0 ] \n",
      "  .. .. .. ..$ gene.loadings.full: num[0 , 0 ] \n",
      "  .. .. .. ..$ sdev              : num(0) \n",
      "  .. .. .. ..$ key               : chr \"ACC\"\n",
      "  .. .. .. ..$ jackstraw         : symbol \u0001NULL\u0001\n",
      "  .. .. .. ..$ misc              : symbol \u0001NULL\u0001\n",
      "  .. ..$ tsne       :Formal class 'dim.reduction' [package \"Seurat\"] with 0 slots\n",
      " Named list()\n",
      "  .. .. .. ..$ cell.embeddings   : num [1:3062, 1:2] 20.8 20.4 23.2 20.3 20.1 ...\n",
      "  .. .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. .. ..$ : chr [1:3062] \"DRF1.1_bcGRPK\" \"DRF1.1_bcCUSZ\" \"DRF1.1_bcCBVN\" \"DRF1.1_bcALFX\" ...\n",
      "  .. .. .. .. .. ..$ : chr [1:2] \"tSNE_1\" \"tSNE_2\"\n",
      "  .. .. .. ..$ gene.loadings     : num[0 , 0 ] \n",
      "  .. .. .. ..$ gene.loadings.full: num[0 , 0 ] \n",
      "  .. .. .. ..$ sdev              : num(0) \n",
      "  .. .. .. ..$ key               : chr \"tSNE_\"\n",
      "  .. .. .. ..$ jackstraw         : symbol \u0001NULL\u0001\n",
      "  .. .. .. ..$ misc              : symbol \u0001NULL\u0001\n",
      "  ..$ assay       : list()\n",
      "  ..$ hvg.info    :'data.frame':\t0 obs. of  0 variables\n",
      "Formal class 'data.frame' [package \"methods\"] with 4 slots\n",
      "  .. .. ..@ .Data    : list()\n",
      "  .. .. ..@ names    : chr(0) \n",
      "  .. .. ..@ row.names: int(0) \n",
      "  .. .. ..@ .S3Class : chr \"data.frame\"\n",
      "  ..$ imputed     :'data.frame':\t0 obs. of  0 variables\n",
      "Formal class 'data.frame' [package \"methods\"] with 4 slots\n",
      "  .. .. ..@ .Data    : list()\n",
      "  .. .. ..@ names    : chr(0) \n",
      "  .. .. ..@ row.names: int(0) \n",
      "  .. .. ..@ .S3Class : chr \"data.frame\"\n",
      "  ..$ cell.names  : chr [1:3062] \"DRF1.1_bcGRPK\" \"DRF1.1_bcCUSZ\" \"DRF1.1_bcCBVN\" \"DRF1.1_bcALFX\" ...\n",
      "  ..$ cluster.tree:List of 1\n",
      "  .. ..$ :List of 4\n",
      "  .. .. ..$ edge       : int [1:32, 1:2] 18 19 19 31 32 32 31 33 33 18 ...\n",
      "  .. .. ..$ edge.length: num [1:32] 36.33 14.77 3.18 6.23 5.35 ...\n",
      "  .. .. ..$ tip.label  : chr [1:17] \"Astrocytes\" \"Differentiating Oligodendrocytes\" \"Dopaminergic Neurons\" \"Endothelial Cells\" ...\n",
      "  .. .. ..$ Nnode      : int 16\n",
      "  .. .. ..- attr(*, \"class\")= chr \"phylo\"\n",
      "  .. .. ..- attr(*, \"order\")= chr \"cladewise\"\n",
      "  ..$ snn         :Formal class 'dgCMatrix' [package \"Matrix\"] with 6 slots\n",
      "  .. .. ..@ i       : int [1:4930919] 0 798 2363 2641 2802 2981 4106 4204 6478 12351 ...\n",
      "  .. .. ..@ p       : int [1:48356] 0 32 81 243 322 455 547 628 726 866 ...\n",
      "  .. .. ..@ Dim     : int [1:2] 48355 48355\n",
      "  .. .. ..@ Dimnames:List of 2\n",
      "  .. .. .. ..$ : chr [1:48355] \"DRF1.1_bcDLGB\" \"DRF1.1_bcHNLM\" \"DRF1.1_bcEAQO\" \"DRF1.1_bcFNKA\" ...\n",
      "  .. .. .. ..$ : chr [1:48355] \"DRF1.1_bcDLGB\" \"DRF1.1_bcHNLM\" \"DRF1.1_bcEAQO\" \"DRF1.1_bcFNKA\" ...\n",
      "  .. .. ..@ x       : num [1:4930919] 1 0.0714 0.0909 0.0714 0.0714 ...\n",
      "  .. .. ..@ factors : list()\n",
      "  ..$ calc.params :List of 10\n",
      "  .. ..$ CreateSeuratObject:List of 13\n",
      "  .. .. ..$ project             : chr \"DRNCellTypes\"\n",
      "  .. .. ..$ min.cells           : num 24\n",
      "  .. .. ..$ min.genes           : num 0\n",
      "  .. .. ..$ is.expr             : num 0\n",
      "  .. .. ..$ normalization.method: NULL\n",
      "  .. .. ..$ scale.factor        : num 10000\n",
      "  .. .. ..$ do.scale            : logi FALSE\n",
      "  .. .. ..$ do.center           : logi FALSE\n",
      "  .. .. ..$ names.field         : num 1\n",
      "  .. .. ..$ names.delim         : chr \"_\"\n",
      "  .. .. ..$ display.progress    : logi TRUE\n",
      "  .. .. ..$ ...                 : symbol \n",
      "  .. .. ..$ time                : POSIXct[1:1], format: \"2018-05-31 11:11:30\"\n",
      "  .. ..$ NormalizeData     :List of 5\n",
      "  .. .. ..$ assay.type          : chr \"RNA\"\n",
      "  .. .. ..$ normalization.method: chr \"LogNormalize\"\n",
      "  .. .. ..$ scale.factor        : num 10000\n",
      "  .. .. ..$ display.progress    : logi TRUE\n",
      "  .. .. ..$ time                : POSIXct[1:1], format: \"2018-05-31 11:11:30\"\n",
      "  .. ..$ ScaleData         :List of 16\n",
      "  .. .. ..$ genes.use         : chr [1:17138] \"X.343C11.2\" \"X0610007P14Rik\" \"X0610009B22Rik\" \"X0610009E02Rik\" ...\n",
      "  .. .. ..$ vars.to.regress   : symbol \n",
      "  .. .. ..$ model.use         : chr \"linear\"\n",
      "  .. .. ..$ use.umi           : logi FALSE\n",
      "  .. .. ..$ do.scale          : logi TRUE\n",
      "  .. .. ..$ do.center         : logi TRUE\n",
      "  .. .. ..$ scale.max         : num 10\n",
      "  .. .. ..$ block.size        : num 1000\n",
      "  .. .. ..$ min.cells.to.block: num 3000\n",
      "  .. .. ..$ display.progress  : logi TRUE\n",
      "  .. .. ..$ assay.type        : chr \"RNA\"\n",
      "  .. .. ..$ do.cpp            : logi TRUE\n",
      "  .. .. ..$ check.for.norm    : logi TRUE\n",
      "  .. .. ..$ do.par            : logi FALSE\n",
      "  .. .. ..$ num.cores         : num 1\n",
      "  .. .. ..$ time              : POSIXct[1:1], format: \"2018-05-31 11:11:35\"\n",
      "  .. ..$ RunCCA            :List of 11\n",
      "  .. .. ..$ group1         : symbol \n",
      "  .. .. ..$ group2         : symbol \n",
      "  .. .. ..$ group.by       : symbol \n",
      "  .. .. ..$ num.cc         : num 50\n",
      "  .. .. ..$ genes.use      : chr [1:3340] \"Hbb.bs\" \"Cartpt\" \"Plp1\" \"Meg3\" ...\n",
      "  .. .. ..$ scale.data     : logi TRUE\n",
      "  .. .. ..$ rescale.groups : logi FALSE\n",
      "  .. .. ..$ ...            :length 1, mode \"...\": (min.cells=*)\n",
      "  .. .. ..$ time           : POSIXct[1:1], format: \"2018-05-31 11:12:56\"\n",
      "  .. .. ..$ object.project : chr \"DRNCellTypes\"\n",
      "  .. .. ..$ object2.project: chr \"DRNCellTypes\"\n",
      "  .. ..$ CalcVarExpRatio   :List of 5\n",
      "  .. .. ..$ reduction.type: chr \"pca\"\n",
      "  .. .. ..$ grouping.var  : chr \"BatchID\"\n",
      "  .. .. ..$ dims.use      : int [1:35] 1 2 3 4 5 6 7 8 9 10 ...\n",
      "  .. .. ..$ verbose       : logi FALSE\n",
      "  .. .. ..$ time          : POSIXct[1:1], format: \"2018-05-31 12:25:23\"\n",
      "  .. ..$ AlignSubspace.cca :List of 9\n",
      "  .. .. ..$ reduction.type    : chr \"cca\"\n",
      "  .. .. ..$ grouping.var      : chr \"BatchID\"\n",
      "  .. .. ..$ dims.align        : int [1:35] 1 2 3 4 5 6 7 8 9 10 ...\n",
      "  .. .. ..$ num.possible.genes: num 2000\n",
      "  .. .. ..$ num.genes         : num 30\n",
      "  .. .. ..$ show.plots        : logi FALSE\n",
      "  .. .. ..$ verbose           : logi FALSE\n",
      "  .. .. ..$ ...               : symbol \n",
      "  .. .. ..$ time              : POSIXct[1:1], format: \"2018-05-31 12:37:50\"\n",
      "  .. ..$ RunTSNE           :List of 13\n",
      "  .. .. ..$ reduction.use  : chr \"cca.aligned\"\n",
      "  .. .. ..$ cells.use      : NULL\n",
      "  .. .. ..$ dims.use       : int [1:35] 1 2 3 4 5 6 7 8 9 10 ...\n",
      "  .. .. ..$ genes.use      : NULL\n",
      "  .. .. ..$ seed.use       : num 1\n",
      "  .. .. ..$ tsne.method    : chr \"Rtsne\"\n",
      "  .. .. ..$ add.iter       : num 0\n",
      "  .. .. ..$ dim.embed      : num 2\n",
      "  .. .. ..$ distance.matrix: NULL\n",
      "  .. .. ..$ reduction.name : chr \"tsne\"\n",
      "  .. .. ..$ reduction.key  : chr \"tSNE_\"\n",
      "  .. .. ..$ ...            :length 2, mode \"...\": (do.fast=*, perplexity=*)\n",
      "  .. .. ..$ time           : POSIXct[1:1], format: \"2018-06-23 17:06:40\"\n",
      "  .. ..$ BuildSNN          :List of 11\n",
      "  .. .. ..$ genes.use     : NULL\n",
      "  .. .. ..$ reduction.type: chr \"cca.aligned\"\n",
      "  .. .. ..$ dims.use      : int [1:35] 1 2 3 4 5 6 7 8 9 10 ...\n",
      "  .. .. ..$ k.param       : num 30\n",
      "  .. .. ..$ plot.SNN      : logi FALSE\n",
      "  .. .. ..$ prune.SNN     : num 0.0667\n",
      "  .. .. ..$ force.recalc  : logi FALSE\n",
      "  .. .. ..$ filename      : NULL\n",
      "  .. .. ..$ save.SNN      : logi TRUE\n",
      "  .. .. ..$ nn.eps        : num 0\n",
      "  .. .. ..$ time          : POSIXct[1:1], format: \"2018-05-31 15:56:29\"\n",
      "  .. ..$ FindClusters.res.3:List of 21\n",
      "  .. .. ..$ genes.use         : NULL\n",
      "  .. .. ..$ reduction.type    : chr \"cca.aligned\"\n",
      "  .. .. ..$ dims.use          : int [1:35] 1 2 3 4 5 6 7 8 9 10 ...\n",
      "  .. .. ..$ k.param           : num 30\n",
      "  .. .. ..$ plot.SNN          : logi FALSE\n",
      "  .. .. ..$ prune.SNN         : num 0.0667\n",
      "  .. .. ..$ print.output      : logi FALSE\n",
      "  .. .. ..$ distance.matrix   : NULL\n",
      "  .. .. ..$ save.SNN          : logi TRUE\n",
      "  .. .. ..$ reuse.SNN         : logi FALSE\n",
      "  .. .. ..$ force.recalc      : logi FALSE\n",
      "  .. .. ..$ nn.eps            : num 0\n",
      "  .. .. ..$ modularity.fxn    : num 1\n",
      "  .. .. ..$ resolution        : num 3\n",
      "  .. .. ..$ algorithm         : num 3\n",
      "  .. .. ..$ n.start           : num 100\n",
      "  .. .. ..$ n.iter            : num 10\n",
      "  .. .. ..$ random.seed       : num 0\n",
      "  .. .. ..$ temp.file.location: NULL\n",
      "  .. .. ..$ edge.file.name    : NULL\n",
      "  .. .. ..$ time              : POSIXct[1:1], format: \"2018-05-31 16:02:12\"\n",
      "  .. ..$ FindClusters.res.2:List of 21\n",
      "  .. .. ..$ genes.use         : NULL\n",
      "  .. .. ..$ reduction.type    : chr \"cca.aligned\"\n",
      "  .. .. ..$ dims.use          : int [1:35] 1 2 3 4 5 6 7 8 9 10 ...\n",
      "  .. .. ..$ k.param           : num 30\n",
      "  .. .. ..$ plot.SNN          : logi FALSE\n",
      "  .. .. ..$ prune.SNN         : num 0.0667\n",
      "  .. .. ..$ print.output      : logi FALSE\n",
      "  .. .. ..$ distance.matrix   : NULL\n",
      "  .. .. ..$ save.SNN          : logi TRUE\n",
      "  .. .. ..$ reuse.SNN         : logi FALSE\n",
      "  .. .. ..$ force.recalc      : logi FALSE\n",
      "  .. .. ..$ nn.eps            : num 0\n",
      "  .. .. ..$ modularity.fxn    : num 1\n",
      "  .. .. ..$ resolution        : num 2\n",
      "  .. .. ..$ algorithm         : num 3\n",
      "  .. .. ..$ n.start           : num 100\n",
      "  .. .. ..$ n.iter            : num 10\n",
      "  .. .. ..$ random.seed       : num 0\n",
      "  .. .. ..$ temp.file.location: NULL\n",
      "  .. .. ..$ edge.file.name    : NULL\n",
      "  .. .. ..$ time              : POSIXct[1:1], format: \"2018-05-31 16:39:31\"\n",
      "  ..$ kmeans      : symbol \u0001NULL\u0001\n",
      "  ..$ spatial     :Formal class 'spatial.info' [package \"Seurat\"] with 0 slots\n",
      " Named list()\n",
      "  .. .. ..$ mix.probs    :'data.frame':\t48375 obs. of  1 variable:\n",
      "  .. .. .. ..$ nGene: int [1:48375] 1114 1300 871 1256 1210 543 600 1427 1115 1036 ...\n",
      "  .. .. ..$ mix.param    :'data.frame':\t0 obs. of  0 variables\n",
      "Formal class 'data.frame' [package \"methods\"] with 4 slots\n",
      "  .. .. .. .. ..@ .Data    : list()\n",
      "  .. .. .. .. ..@ names    : chr(0) \n",
      "  .. .. .. .. ..@ row.names: int(0) \n",
      "  .. .. .. .. ..@ .S3Class : chr \"data.frame\"\n",
      "  .. .. ..$ final.prob   :'data.frame':\t0 obs. of  0 variables\n",
      "Formal class 'data.frame' [package \"methods\"] with 4 slots\n",
      "  .. .. .. .. ..@ .Data    : list()\n",
      "  .. .. .. .. ..@ names    : chr(0) \n",
      "  .. .. .. .. ..@ row.names: int(0) \n",
      "  .. .. .. .. ..@ .S3Class : chr \"data.frame\"\n",
      "  .. .. ..$ insitu.matrix:'data.frame':\t0 obs. of  0 variables\n",
      "Formal class 'data.frame' [package \"methods\"] with 4 slots\n",
      "  .. .. .. .. ..@ .Data    : list()\n",
      "  .. .. .. .. ..@ names    : chr(0) \n",
      "  .. .. .. .. ..@ row.names: int(0) \n",
      "  .. .. .. .. ..@ .S3Class : chr \"data.frame\"\n",
      "  ..$ misc        : symbol \u0001NULL\u0001\n",
      "  ..$ version     :Classes 'package_version', 'numeric_version'  hidden list of 1\n",
      "  .. ..$ : int [1:3] 2 3 1\n"
     ]
    }
   ],
   "source": [
    "str(rds_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "6c912839-ef58-4017-955d-4e760863fb43",
   "metadata": {},
   "outputs": [],
   "source": [
    "hdf5_file <- Read10X_h5(filename = \"C:/Users/nidhi/Downloads/20k_PBMC_3p_HT_nextgem_Chromium_X_filtered_feature_bc_matrix.h5\", use.names = TRUE, unique.features = TRUE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "03db1bea-7aae-4037-b661-28013cc2c248",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "10 x 3 sparse Matrix of class \"dgCMatrix\"\n",
       "            AAACCCAAGAGCAGCT-1 AAACCCACAATACCTG-1 AAACCCACACAACCGC-1\n",
       "MIR1302-2HG                  .                  .                  .\n",
       "FAM138A                      .                  .                  .\n",
       "OR4F5                        .                  .                  .\n",
       "AL627309.1                   .                  .                  .\n",
       "AL627309.3                   .                  .                  .\n",
       "AL627309.2                   .                  .                  .\n",
       "AL627309.5                   .                  .                  .\n",
       "AL627309.4                   .                  .                  .\n",
       "AP006222.2                   .                  .                  .\n",
       "AL732372.1                   .                  .                  ."
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "hdf5_file[0:10,0:3]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "82713a77-f34c-40ec-87a0-9cb69b91b06a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Formal class 'Seurat' [package \"SeuratObject\"] with 13 slots\n",
      "  ..@ assays      :List of 1\n",
      "  .. ..$ RNA:Formal class 'Assay5' [package \"SeuratObject\"] with 8 slots\n",
      "  .. .. .. ..@ layers    :List of 1\n",
      "  .. .. .. .. ..$ counts:Formal class 'dgCMatrix' [package \"Matrix\"] with 6 slots\n",
      "  .. .. .. .. .. .. ..@ i       : int [1:54878285] 16 25 30 36 48 59 62 85 86 97 ...\n",
      "  .. .. .. .. .. .. ..@ p       : int [1:23838] 0 3085 4930 7867 12721 13128 16392 19450 21279 23327 ...\n",
      "  .. .. .. .. .. .. ..@ Dim     : int [1:2] 36601 23837\n",
      "  .. .. .. .. .. .. ..@ Dimnames:List of 2\n",
      "  .. .. .. .. .. .. .. ..$ : NULL\n",
      "  .. .. .. .. .. .. .. ..$ : NULL\n",
      "  .. .. .. .. .. .. ..@ x       : num [1:54878285] 1 1 3 1 1 1 1 2 1 1 ...\n",
      "  .. .. .. .. .. .. ..@ factors : list()\n",
      "  .. .. .. ..@ cells     :Formal class 'LogMap' [package \"SeuratObject\"] with 1 slot\n",
      "  .. .. .. .. .. ..@ .Data: logi [1:23837, 1] TRUE TRUE TRUE TRUE TRUE TRUE ...\n",
      "  .. .. .. .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. .. .. .. ..$ : chr [1:23837] \"AAACCCAAGAGCAGCT-1\" \"AAACCCACAATACCTG-1\" \"AAACCCACACAACCGC-1\" \"AAACCCACACACAGAG-1\" ...\n",
      "  .. .. .. .. .. .. .. ..$ : chr \"counts\"\n",
      "  .. .. .. .. .. ..$ dim     : int [1:2] 23837 1\n",
      "  .. .. .. .. .. ..$ dimnames:List of 2\n",
      "  .. .. .. .. .. .. ..$ : chr [1:23837] \"AAACCCAAGAGCAGCT-1\" \"AAACCCACAATACCTG-1\" \"AAACCCACACAACCGC-1\" \"AAACCCACACACAGAG-1\" ...\n",
      "  .. .. .. .. .. .. ..$ : chr \"counts\"\n",
      "  .. .. .. ..@ features  :Formal class 'LogMap' [package \"SeuratObject\"] with 1 slot\n",
      "  .. .. .. .. .. ..@ .Data: logi [1:36601, 1] TRUE TRUE TRUE TRUE TRUE TRUE ...\n",
      "  .. .. .. .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. .. .. .. ..$ : chr [1:36601] \"MIR1302-2HG\" \"FAM138A\" \"OR4F5\" \"AL627309.1\" ...\n",
      "  .. .. .. .. .. .. .. ..$ : chr \"counts\"\n",
      "  .. .. .. .. .. ..$ dim     : int [1:2] 36601 1\n",
      "  .. .. .. .. .. ..$ dimnames:List of 2\n",
      "  .. .. .. .. .. .. ..$ : chr [1:36601] \"MIR1302-2HG\" \"FAM138A\" \"OR4F5\" \"AL627309.1\" ...\n",
      "  .. .. .. .. .. .. ..$ : chr \"counts\"\n",
      "  .. .. .. ..@ default   : int 1\n",
      "  .. .. .. ..@ assay.orig: chr(0) \n",
      "  .. .. .. ..@ meta.data :'data.frame':\t36601 obs. of  0 variables\n",
      "  .. .. .. ..@ misc      :List of 1\n",
      "  .. .. .. .. ..$ calcN: logi TRUE\n",
      "  .. .. .. ..@ key       : chr \"rna_\"\n",
      "  ..@ meta.data   :'data.frame':\t23837 obs. of  3 variables:\n",
      "  .. ..$ orig.ident  : Factor w/ 1 level \"SeuratProject\": 1 1 1 1 1 1 1 1 1 1 ...\n",
      "  .. ..$ nCount_RNA  : num [1:23837] 9203 5705 8877 22005 626 ...\n",
      "  .. ..$ nFeature_RNA: int [1:23837] 3085 1845 2937 4854 407 3264 3058 1829 2048 3485 ...\n",
      "  ..@ active.assay: chr \"RNA\"\n",
      "  ..@ active.ident: Factor w/ 1 level \"SeuratProject\": 1 1 1 1 1 1 1 1 1 1 ...\n",
      "  .. ..- attr(*, \"names\")= chr [1:23837] \"AAACCCAAGAGCAGCT-1\" \"AAACCCACAATACCTG-1\" \"AAACCCACACAACCGC-1\" \"AAACCCACACACAGAG-1\" ...\n",
      "  ..@ graphs      : list()\n",
      "  ..@ neighbors   : list()\n",
      "  ..@ reductions  : list()\n",
      "  ..@ images      : list()\n",
      "  ..@ project.name: chr \"SeuratProject\"\n",
      "  ..@ misc        : list()\n",
      "  ..@ version     :Classes 'package_version', 'numeric_version'  hidden list of 1\n",
      "  .. ..$ : int [1:3] 5 0 1\n",
      "  ..@ commands    : list()\n",
      "  ..@ tools       : list()\n"
     ]
    }
   ],
   "source": [
    "seurat_obj_hdf5 <- CreateSeuratObject(counts = hdf5_file)\n",
    "str(seurat_obj_hdf5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "a0dd4283-fb9b-4f10-ab4b-47576ec2c4aa",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "10 x 3 sparse Matrix of class \"dgCMatrix\"\n",
       "            AAACCCAAGAAACACT-1 AAACCCAAGAAACCAT-1 AAACCCAAGAAACCCA-1\n",
       "MIR1302-2HG                  .                  .                  .\n",
       "FAM138A                      .                  .                  .\n",
       "OR4F5                        .                  .                  .\n",
       "AL627309.1                   .                  .                  .\n",
       "AL627309.3                   .                  .                  .\n",
       "AL627309.2                   .                  .                  .\n",
       "AL627309.5                   .                  .                  .\n",
       "AL627309.4                   .                  .                  .\n",
       "AP006222.2                   .                  .                  .\n",
       "AL732372.1                   .                  .                  ."
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "count_matrix <- ReadMtx(mtx=\"C:/Users/nidhi/Downloads/20k_PBMC_3p_HT_nextgem_Chromium_X_raw_feature_bc_matrix/raw_feature_bc_matrix/matrix.mtx.gz\",features=\"C:/Users/nidhi/Downloads/20k_PBMC_3p_HT_nextgem_Chromium_X_raw_feature_bc_matrix/raw_feature_bc_matrix/features.tsv.gz\", cells=\"C:/Users/nidhi/Downloads/20k_PBMC_3p_HT_nextgem_Chromium_X_raw_feature_bc_matrix/raw_feature_bc_matrix/barcodes.tsv.gz\")\n",
    "count_matrix[0:10,0:3]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "d097b9a0-e7bb-4b5c-8349-8d5e51b4022c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Formal class 'Seurat' [package \"SeuratObject\"] with 13 slots\n",
      "  ..@ assays      :List of 1\n",
      "  .. ..$ RNA:Formal class 'Assay5' [package \"SeuratObject\"] with 8 slots\n",
      "  .. .. .. ..@ layers    :List of 1\n",
      "  .. .. .. .. ..$ counts:Formal class 'dgCMatrix' [package \"Matrix\"] with 6 slots\n",
      "  .. .. .. .. .. .. ..@ i       : int [1:62927657] 16186 28979 36562 31495 12196 13970 29828 11767 13111 25486 ...\n",
      "  .. .. .. .. .. .. ..@ p       : int [1:3069479] 0 1 1 1 1 1 3 4 4 4 ...\n",
      "  .. .. .. .. .. .. ..@ Dim     : int [1:2] 36601 3069478\n",
      "  .. .. .. .. .. .. ..@ Dimnames:List of 2\n",
      "  .. .. .. .. .. .. .. ..$ : NULL\n",
      "  .. .. .. .. .. .. .. ..$ : NULL\n",
      "  .. .. .. .. .. .. ..@ x       : num [1:62927657] 1 1 1 1 1 1 1 1 1 1 ...\n",
      "  .. .. .. .. .. .. ..@ factors : list()\n",
      "  .. .. .. ..@ cells     :Formal class 'LogMap' [package \"SeuratObject\"] with 1 slot\n",
      "  .. .. .. .. .. ..@ .Data: logi [1:3069478, 1] TRUE TRUE TRUE TRUE TRUE TRUE ...\n",
      "  .. .. .. .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. .. .. .. ..$ : chr [1:3069478] \"AAACCCAAGAAACACT-1\" \"AAACCCAAGAAACCAT-1\" \"AAACCCAAGAAACCCA-1\" \"AAACCCAAGAAACCTG-1\" ...\n",
      "  .. .. .. .. .. .. .. ..$ : chr \"counts\"\n",
      "  .. .. .. .. .. ..$ dim     : int [1:2] 3069478 1\n",
      "  .. .. .. .. .. ..$ dimnames:List of 2\n",
      "  .. .. .. .. .. .. ..$ : chr [1:3069478] \"AAACCCAAGAAACACT-1\" \"AAACCCAAGAAACCAT-1\" \"AAACCCAAGAAACCCA-1\" \"AAACCCAAGAAACCTG-1\" ...\n",
      "  .. .. .. .. .. .. ..$ : chr \"counts\"\n",
      "  .. .. .. ..@ features  :Formal class 'LogMap' [package \"SeuratObject\"] with 1 slot\n",
      "  .. .. .. .. .. ..@ .Data: logi [1:36601, 1] TRUE TRUE TRUE TRUE TRUE TRUE ...\n",
      "  .. .. .. .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. .. .. .. ..$ : chr [1:36601] \"MIR1302-2HG\" \"FAM138A\" \"OR4F5\" \"AL627309.1\" ...\n",
      "  .. .. .. .. .. .. .. ..$ : chr \"counts\"\n",
      "  .. .. .. .. .. ..$ dim     : int [1:2] 36601 1\n",
      "  .. .. .. .. .. ..$ dimnames:List of 2\n",
      "  .. .. .. .. .. .. ..$ : chr [1:36601] \"MIR1302-2HG\" \"FAM138A\" \"OR4F5\" \"AL627309.1\" ...\n",
      "  .. .. .. .. .. .. ..$ : chr \"counts\"\n",
      "  .. .. .. ..@ default   : int 1\n",
      "  .. .. .. ..@ assay.orig: chr(0) \n",
      "  .. .. .. ..@ meta.data :'data.frame':\t36601 obs. of  0 variables\n",
      "  .. .. .. ..@ misc      :List of 1\n",
      "  .. .. .. .. ..$ calcN: logi TRUE\n",
      "  .. .. .. ..@ key       : chr \"rna_\"\n",
      "  ..@ meta.data   :'data.frame':\t3069478 obs. of  3 variables:\n",
      "  .. ..$ orig.ident  : Factor w/ 1 level \"SeuratProject\": 1 1 1 1 1 1 1 1 1 1 ...\n",
      "  .. ..$ nCount_RNA  : num [1:3069478] 1 0 0 0 0 2 1 0 0 3 ...\n",
      "  .. ..$ nFeature_RNA: int [1:3069478] 1 0 0 0 0 2 1 0 0 3 ...\n",
      "  ..@ active.assay: chr \"RNA\"\n",
      "  ..@ active.ident: Factor w/ 1 level \"SeuratProject\": 1 1 1 1 1 1 1 1 1 1 ...\n",
      "  .. ..- attr(*, \"names\")= chr [1:3069478] \"AAACCCAAGAAACACT-1\" \"AAACCCAAGAAACCAT-1\" \"AAACCCAAGAAACCCA-1\" \"AAACCCAAGAAACCTG-1\" ...\n",
      "  ..@ graphs      : list()\n",
      "  ..@ neighbors   : list()\n",
      "  ..@ reductions  : list()\n",
      "  ..@ images      : list()\n",
      "  ..@ project.name: chr \"SeuratProject\"\n",
      "  ..@ misc        : list()\n",
      "  ..@ version     :Classes 'package_version', 'numeric_version'  hidden list of 1\n",
      "  .. ..$ : int [1:3] 5 0 1\n",
      "  ..@ commands    : list()\n",
      "  ..@ tools       : list()\n"
     ]
    }
   ],
   "source": [
    "seurat_obj_mtx <- CreateSeuratObject(counts = count_matrix)\n",
    "str(seurat_obj_mtx)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4c95c516-11e0-4cfb-8a0f-1be1126b6062",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.1.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
